File: leetcode/101-200/148-sort-list/sort_list.py
"""LeetCode 148 — Sort List (Python mirror / oracle).

Top-down merge sort over a singly linked list. The Kāra version models nodes as
Vec-owned with a `weak` next overlay and rewires the weak links in place; here we
mirror the same algorithm on index links (`next_of[i]`) so the emitted order is
identical.
"""


def sort_list(vals):
    n = len(vals)
    next_of = [-1] * n
    for i in range(n - 1):
        next_of[i] = i + 1

    def split_mid(head):
        slow = head
        fast = next_of[head]
        while fast != -1:
            fast = next_of[fast]
            if fast != -1:
                slow = next_of[slow]
                fast = next_of[fast]
        mid = next_of[slow]
        next_of[slow] = -1
        return mid

    def merge(a, b):
        head = tail = -1
        while a != -1 and b != -1:
            if vals[a] <= vals[b]:
                pick, a = a, next_of[a]
            else:
                pick, b = b, next_of[b]
            if head == -1:
                head = pick
            else:
                next_of[tail] = pick
            tail = pick
        rest = a if a != -1 else b
        if tail != -1:
            next_of[tail] = rest
        return head

    def sort_chain(head):
        if head == -1 or next_of[head] == -1:
            return head
        mid = split_mid(head)
        left = sort_chain(head)
        right = sort_chain(mid)
        return merge(left, right)

    out = []
    if n == 0:
        return out
    cur = sort_chain(0)
    while cur != -1:
        out.append(vals[cur])
        cur = next_of[cur]
    return out


def run(vals):
    print(" ".join(str(x) for x in sort_list(vals)))


def main():
    run([4, 2, 1, 3])
    run([-1, 5, 3, 4, 0])
    run([])
    run([1])
    run([3, 3, 1, 2, 2])
    run([5, 4, 3, 2, 1])


main()

File: leetcode/101-200/148-sort-list/bench/sort_list.c
#include <stdio.h>
#include <stdlib.h>

static long split_mid(long *nxt, long head) {
    long slow = head;
    long fast = nxt[head];
    while (fast != -1) {
        fast = nxt[fast];
        if (fast != -1) {
            slow = nxt[slow];
            fast = nxt[fast];
        }
    }
    long mid = nxt[slow];
    nxt[slow] = -1;
    return mid;
}

static long merge(const long *val, long *nxt, long a, long b) {
    long ai = a, bi = b, head = -1, tail = -1;
    while (ai != -1 && bi != -1) {
        if (val[ai] <= val[bi]) {
            if (head == -1) head = ai; else nxt[tail] = ai;
            tail = ai;
            ai = nxt[ai];
        } else {
            if (head == -1) head = bi; else nxt[tail] = bi;
            tail = bi;
            bi = nxt[bi];
        }
    }
    long rest = (ai != -1) ? ai : bi;
    if (tail != -1) {
        if (rest == -1) nxt[tail] = -1; else nxt[tail] = rest;
    }
    return head;
}

static long sort_chain(const long *val, long *nxt, long head) {
    if (head == -1) return -1;
    if (nxt[head] == -1) return head;
    long mid = split_mid(nxt, head);
    long left = sort_chain(val, nxt, head);
    long right = sort_chain(val, nxt, mid);
    return merge(val, nxt, left, right);
}

int main(void) {
    long n = 20000, passes = 180, vr = 100000;
    long *val = malloc(n * sizeof(long));
    long *nxt = malloc(n * sizeof(long));
    long state = 12345;
    for (long i = 0; i < n; i++) {
        state = (state * 1103515245L + 12345L) & 2147483647L;
        val[i] = state % vr;
        nxt[i] = -1;
    }

    long sink = 0;
    for (long p = 0; p < passes; p++) {
        state = (state * 1103515245L + 12345L) & 2147483647L;
        long idx = state % n;
        state = (state * 1103515245L + 12345L) & 2147483647L;
        val[idx] = state % vr;

        for (long i = 0; i < n - 1; i++) nxt[i] = i + 1;
        nxt[n - 1] = -1;

        long head = sort_chain(val, nxt, 0);

        long cur = head;
        long pos = 1;
        while (cur != -1) {
            sink += pos * val[cur];
            pos++;
            cur = nxt[cur];
        }
    }
    printf("%ld\n", sink);
    free(val);
    free(nxt);
    return 0;
}
